libraries/tock-cells/src/take_cell.rs

```diff
@@ -13,29 +12,32 @@ use core::ptr;
 /// operate on a borrow within a closure. Attempts to take the value from inside
 /// a `TakeCell` may fail by returning `None`.
 pub struct TakeCell<'a, T: 'a + ?Sized> {
-    val: UnsafeCell<Option<&'a mut T>>,
+    val: Cell<Option<&'a mut T>>,
 }
 
 impl<'a, T: ?Sized> TakeCell<'a, T> {
     pub const fn empty() -> TakeCell<'a, T> {
         TakeCell {
-            val: UnsafeCell::new(None),
+            val: Cell::new(None),
         }
     }
 
     /// Creates a new `TakeCell` containing `value`
     pub fn new(value: &'a mut T) -> TakeCell<'a, T> {
         TakeCell {
-            val: UnsafeCell::new(Some(value)),
+            val: Cell::new(Some(value)),
         }
     }
 
     pub fn is_none(&self) -> bool {
-        unsafe { (&*self.val.get()).is_none() }
+        let inner = self.take();
+        let return_val = inner.is_none();
+        self.val.set(inner);
+        return_val
```

Comment:
At one point we were somewhat deeply concerned with the overhead and performance of cells ... will the above successfully be reduced by the compiler to:
```rust
    pub fn is_none(&self) -> bool {
        unsafe {
            self.val.as_ptr().as_ref().is_none()
        }
    }
```
or is there an overhead to moving references in and out we should be concerned about? (Or is this a needless micro-optimization question?)

Reply:
In general, there's a reason why one can't extract a `&U` from a `Cell<U>`: because there can be another reference to the same `Cell`, that could concurrently call `set`, thereby violating the aliasing rules (i.e. while accessing a `&U`, the same object is being concurrently mutated).

In this case, what if the inner `Option` is changed from `None` to `Some(foo)` while we're calling `is_none`? Can you prove that this is safe? Unless proven otherwise, writing new contents to an `Option` isn't an atomic operation, so the `Option` may be in an invalid state while the `is_none` method is called.

Also, has the potential overhead of this take-and-set pattern been benchmarked? It may well be that this is all optimized by the compiler, and that the safe version has no noticeable overhead. I'd err on the side of less `unsafe` code unless proven necessary by benchmarks (and proven correct).

Reply:
Perhaps I am confused as to the use of `TakeCell` and/or Rust semantics (I'm still fairly new to both the language and the project), but from what I understand `TakeCell` cannot be used in multiple threads at the same time because it does not implement `Sync`. Therefore, there should not be a case where a change to the interior state can happen concurrently with another access to the same `TakeCell`.

I am working to benchmark the overhead, and hopefully show that the Rust compiler is able to optimize this take-and-set pattern away. I should have that data by the end of next week, if not sooner.

Reply:
> Perhaps I am confused as to the use of `TakeCell` and/or Rust semantics (I'm still fairly new to both the language and the project), but from what I understand `TakeCell` cannot be used in multiple threads at the same time because it does not implement `Sync`. Therefore, there should not be a case where a change to the interior state can happen concurrently with another access to the same `TakeCell`.

Note that multi-threading is not the only scenario where Rust's aliasing rules are useful. For example, if you have a container and iterators pointing to the container at the same time (in a single thread), modifying the container can invalidate the iterators (that's why you typically cannot modify the container in Rust).

I'm not saying that `TakeCell` isn't safe in a single-thread scenario (although interruptions can somewhat introduce some amount of "concurrency"), but I'm just cautious about any `unsafe` block without a proof that it's safe. IMO every use of the `unsafe` keyword should be preceded by a comment explaining why it's safe, and why an alternative no-`unsafe` implementation isn't suitable.

Reply:
Analyzing the output assembly code shows that the Rust compiler optimizations are in fact smart enough to recognize this take-and-set pattern as a noop and optimize it away, as if it were just checking the interior `Option`. There is no overhead incurred at runtime.